### ISSUE TITLE
Environment Variables keeps reading from .env

### ISSUE THREAD

**Issue by my-name-is-nheo** [react-native-config/react-native-config#652]
running `./gradlew bundleRelease` to generate app bundles to release to google play store. 

Debug mode reads correct environment variants with respective .env files. However, all variant's app bundles are being generated with only the .env file. 

Root 
![image](https://user-images.githubusercontent.com/58013260/160726855-f41ea95a-3d93-4924-8c71-1b65cef58a0f.png)

app/build.gradle
![image](https://user-images.githubusercontent.com/58013260/160727040-cbec5ad8-a05d-4265-a787-a632044013d1.png)

![image](https://user-images.githubusercontent.com/58013260/160727183-16596b18-0122-4fee-bfcf-f32a6c7ee1ae.png)

1. tried adding proguard override `-keep class com.dottid.BuildConfig { *; }`
2. tried `ENVFILE=.env.prod ./gradlew bundleRelease`, but getting a bunch of gradle errors, `... without declaring an explicit or implicit dependency`
3. tried changing name of `.env` to `.env.developer` but then getting a missing env error when i run command `./gradlew bundleRelease`

Thank you for your time

**Comment by jjdp** [comment]
are you using monterey?

**Comment by fahry-mohammed** [comment]
> monterey

I am using Bumblebee but get the same error. without .env file, build fails.

**Comment by justin-tay** [comment]
The specific issue from the topic starter is likely because he is building using `./gradlew bundleRelease` to build all the release variants at one go instead of specifically specifying the full variant eg. `./gradlew bundleProductionRelease`.

You can look at https://github.com/luggit/react-native-config/blob/master/android/dotenv.gradle in this repository to see the logic for how the env file is determined. From the logic I don't think it is possible for it to determine the correct env file to use if you are building more than one build variant at a time. In which case it will fallback on .env.

As a sanity check you can after building see `android/app/build/generated/source/buildConfig/{productFlavor}/{buildType}/{package}/BuildConfig.java` to check if the values are as expected.

**Comment by justin-tay** [comment]
> I just updated Android studio & command line tools. Since then, I'm not receiving the the ENVFILE variable (with `System.env['ENVFILE']`) when dotenv.gradle is executed (Checked by printout output of `System.env` in `dotenv.gradle`).
> 
> Could there be a change in the new version of either android studio or command line tools which affects the ability to read this variable?

Just tried. Using Android Studio Chipmunk on Windows and tested with `SET ENVFILE=.env.staging && react-native run-android` in `package.json` and it works. That said I made sure that I didn't have any pre-existing gradle daemons running. There might be potential issues with a pre-existing gradle daemon and reading the current environment variables. I actually don't use the ENVFILE environment variable and just configure variants for Android and schemes for iOS.

**Comment by Mr0cket** [comment]
> That said I made sure that I didn't have any pre-existing gradle daemons running. There might be potential issues with a pre-existing gradle daemon and reading the current environment variables.

Good point. After retrying without pre-existing daemon running, the `ENVFILE` variable was read correctly.

This may seem pretty common sense in retrospect, but might still be worth adding a small note to the README for anyone who encounters the same issue.

**Comment by santitopo** [comment]
Similar thing to what @Mr0cket reports happening here. The `ENVFILE=.env.something` way of setting the env file seems to have stopped working as expected.

Up till now I have had different npm scripts, each doing `ENVFILE=.env.something react-native run-android`, and the envfile would be correctly changed if I ran two different builds in a row. 

Now the first run will pick the correct env file, and the ones following will use the same ENVFILE variable, unless the gradle daemon is stopped with `./gradlew --stop` before the build.

Not sure if it can have something to do but I've recently updated my JDK version from JDK 8 to 11 (Zulu OpenJDK as specified in the latest RN docs) and also the Xcode Command Line Tools to 13.2.0.

I'm running in Mac M1
`react-native: 0.67.2`
`react-native-config: 1.4.6`
`macOS: BigSur (11.6.2)`

**Comment by vinhtnk** [comment]
my solution is `cp .env.prod .env` before run a script, and .env was ignored from git

**Comment by Stas-Buzunko** [comment]
> my solution is `cp .env.prod .env` before run a script, and .env was ignored from git

only this worked.
project.ext.envConfigFiles and ENVFILE=.env.something are ignored

"react-native-config": "^1.4.6"

**Comment by chamaldesilva** [comment]
> ./gradlew --stop

This one worked for me finally!

**Comment by BoilingOil** [comment]
right above `project.ext.envConfigFiles`, set this:
```
project.ext.defaultEnvFile ="../path/to/.env.file"
```

**Comment by ToshKoevoets** [comment]
Yes super annoying, seems to be an issue with m1 or montery keeping gradle deamon in memory. For fastlane added the -- stop command in the build script

    gradle(
      task: "--stop"
    )

**Comment by anees17861** [comment]
Hi,

I had run into the same issue and did the following approach.

Instead of providing envConfigFiles at top level, what I did was provide default env file in the product flavor itself

```
    flavorDimensions "phase"
    productFlavors {
        dev {
            dimension "phase"
            project.ext.defaultEnvFile = ".env.dev"
            apply from: project(':react-native-config').projectDir.getPath() + "/dotenv.gradle"
            applicationId project.env.get("APP_ID")
        }
        stage {
            dimension "phase"
            project.ext.defaultEnvFile = ".env.stage"
            apply from: project(':react-native-config').projectDir.getPath() + "/dotenv.gradle"
            applicationId project.env.get("APP_ID")
        }
        prod {
            dimension "phase"
            project.ext.defaultEnvFile = ".env.prod"
            apply from: project(':react-native-config').projectDir.getPath() + "/dotenv.gradle"
            applicationId project.env.get("APP_ID")
        }
    }
```
This should allow to build multiple variants with one command. 

Since `apply` is getting invoked multiple times there might be a performance hit. Maybe someone with gradle experience can comment here
One drawback here is that env variable should be used only within flavor definition. Once it goes outside the block, it may be overwritten by the next flavor.
Also more complex combination of flavors may become difficult to handle.

**Comment by AdamGerthel** [comment]
I just stumbled on this as well... the strange this is that it's only affecting one of my projects, not both. `./gradlew --stop` worked for me so I've added it to my fastlane as per [#issuecomment-1230403293](https://github.com/luggit/react-native-config/issues/652#issuecomment-1230403293)

**Comment by CaioSantos** [comment]
> right above `project.ext.envConfigFiles`, set this:
> 
> ```
> project.ext.defaultEnvFile ="../path/to/.env.file"
> ```

This one line worked like a charm in my case

**Comment by ToshKoevoets** [comment]
When running it locally the java process also stays running forever on the m1 new mac books, taking quite some memory, so works better to stop the process. 

gradle(
  task: "--stop"
)

**Comment by Beka-Bat1** [comment]
from your root folder try this

`cd android && ./gradlew clean && ENVFILE=../.env.production ./gradlew bundleRelease && cd ..
`

**Comment by personalnadir** [comment]
> my solution is `cp .env.prod .env` before run a script, and .env was ignored from git

I have a few apps using the same codebase and was attempting to keep the .env files in separate folders. Using the ENVFILE=app1/.env.development command would not work for me at all. Only the above has worked correctly

**Comment by mahardikaindra** [comment]
> > my solution is `cp .env.prod .env` before run a script, and .env was ignored from git
> 
> only this worked. project.ext.envConfigFiles and ENVFILE=.env.something are ignored
> 
> "react-native-config": "^1.4.6"
> 
> "./gradlew --stop" - this helped to reset to .env from .env.beta (no idea how it .env.beta was selected before 🤨)

Yes, worked for me

**Comment by Abhijeetk28** [comment]
Hi , I am still facing this issue !
The release builds always picks up .env config instead of any other extenstion.

**Comment by viet795** [comment]
add defaultEnvFile=.env.prod in to file gradle.properties

**Comment by ashcode1** [comment]
In my case this issue was caused by calling 
`apply from: project(':react-native-config').projectDir.getPath() + "/dotenv.gradle"
`
too early. It must be called AFTER: 
`project.ext.envConfigFiles = [
    debug: ".env.development",
    release: ".env.production",
    anothercustombuild: ".env",
]`
